### ISSUE TITLE
Explain cacheKeyForTree rules for addon deduplication

### ISSUE THREAD

**Issue by mydea** [ember-engines/ember-engines#606]
I just spent a rather long time trying to figure out why our internal, shared addon code was included in all lazy loaded engine-vendor.js files. Turns out, this was caused by us defining a custom `treeForAddon()` hook, which lead to `cacheKeyForTree(addon)` returning null, which lead to the addon not being deduplicated.

Apart from the fact that this is IMHO a bit counterintuitive (what is the benefit of actually including the addon code again in the engine, as only one addon version can be used anyhow at any time? What would a scenario be where this behavior is useful/desired?), it is not documented at all (as far as I've seen).

At the very least, it would be great to write a paragraph or similar about how addon deduplication works, what the rules are, and about these possible gotchas (e.g. "If an addon provides a custom `treeForAddon` hook, it will need to also overwrite `cacheKeyForTree` accordingly to ensure the addon is not included multiple times"). I could write something up, but not entirely sure where to put it.

**Comment by nmcclay** [comment]
This is interesting, because https://github.com/ember-engines/ember-engines/issues/502 is actually about the exact opposite use case.  There is an awful lot of confusion around addons and engines it seems and not a lot of consensus on the best approach from what I have learned here.

**Comment by mydea** [comment]
Sure, there is a more complex case where each engine might have a different version of an addon - how to solve this, is a bit more tricky. But in our case, the host app, engines & shared addon all live in a yarn workspace, so they all have exactly the same version, and the addon is included in the host app as well as in all engines - so it is pretty unnecessary and unexpected that the engine addon code is duplicated in every engine.

**Comment by mydea** [comment]
Not sure if that is the same issue IMHO - the issue you describe is generally hard to solve, I guess. All possible cases have down sides here:

a) Include shared deps into the host app - Downside: Increases initial bundle size
b) Include shared deps into each engine bundle - Downside: Duplicated stuff in multiple engine bundles
c) Somehow split out shared deps into dedicated bundles (e.g. `/engine-assets/shared-by-a-and-b.js`) - Downside: You might end up with 20 bundles (e.g. shared-by-a-and-b, shared-by-b-and-c-, shared-by-a-b-and-c, ...), and it is probably pretty complicated from a built step perspective

(I think b) is the currently used approach, but might be mistaken)

However, this specific issue here is about the more straightforward issue where an addon is included in the host app _and_ an engine.

**Comment by dgeb** [comment]
> (I think b) is the currently used approach, but might be mistaken)

Yes, this is correct.

> However, this specific issue here is about the more straightforward issue where an addon is included in the host app and an engine.

@mydea I agree the subject of this issue is much more straightforward to identify and address. Docs would be helpful but it would be even more helpful to alert devs of this potential surprise at build time. Otherwise, it would be very easy for this kind of duplication to go unnoticed.

**Comment by villander** [comment]
@dgeb we can put a section on [ember-engines.com](http://ember-engines.com/guide) like "Addons and Dependencies" as Ember Guides.

@mydea I can hep you write this, if you need

**Comment by villander** [comment]
@dgeb what can we do to remove duplicate dependencies? do I need to do an RFC or should this [pull request](#595) resolve?

@mydea the approach `b` is solved with #595 ?

**Comment by mydea** [comment]
Yes, b) is supposed to work - I think #595 basically fixes "bugs" with that approach for nested dependencies. However, it still leaves the `cacheKeyForTree` issue unresolved.

I could champion some engine documentation work, if someone could guide me to where/how to proceed - also see https://github.com/ember-engines/ember-engines.com/issues/16 and #540 , the state of engine documentation is a bit unclear to me right now.

**Comment by villander** [comment]
> At the very least, it would be great to write a paragraph or similar about how addon deduplication works, what the rules are, and about these possible gotchas (e.g. "If an addon provides a custom treeForAddon hook, it will need to also overwrite cacheKeyForTree accordingly to ensure the addon is not included multiple times"). I could write something up, but not entirely sure where to put it.

@mydea you can put this on "Addons and Dependencies" on `ember-engines.com`.  You can create this section if you need. And if you can make a repo production to reproduce this bug? and I'll work on this.

> I could champion some engine documentation work, if someone could guide me to where/how to proceed - also see ember-engines/ember-engines.com#16 and #540 , the state of engine documentation is a bit unclear to me right now.

About this, I can help on new documentation, but about dependencies I need help of @dgeb because many things related with dependencies don't clear for me as well

**Comment by 2hu12** [comment]
> @mydea the approach `b` is solved with #595 ?

#595 aims for a narrower situation which is https://github.com/ember-engines/ember-engines/pull/356#issuecomment-288571816.

This "Shared addon deeper than engine" deduplication was just not implemented at that time and it already had relative tests but skiped. What #595 does was implementing this feature and unskip the tests.
